src/main/java/org/sonarlint/intellij/core/ServerIssueUpdater.java

```diff
@@ -0,0 +1,130 @@
+
+  private static final int THREADS_NUM = 5;
+
+  private ExecutorService executorService;
+
+  private final IssueStore store;
+
+  public ServerIssueUpdater(Project project, IssueStore store) {
+    super(project);
+    this.store = store;
+  }
+
+  public void trackServerIssues(Set<VirtualFile> virtualFiles) {
+    virtualFiles.forEach(this::trackServerIssues);
+  }
+
+  private void trackServerIssues(VirtualFile virtualFile) {
+    SonarLintProjectSettings projectSettings = SonarLintUtils.get(myProject, SonarLintProjectSettings.class);
+    if (!projectSettings.isBindingEnabled()) {
+      // not in connected mode
+      return;
+    }
+
+    String serverId = projectSettings.getServerId();
+    String moduleKey = projectSettings.getProjectKey();
+    if (serverId == null || moduleKey == null) {
+      // not bound to SQ project
+      return;
+    }
+
+    ConnectedSonarLintEngine engine = SonarLintUtils.get(SonarLintServerManager.class).getConnectedEngine(serverId);
+
+    String relativePath = getRelativePath(virtualFile);
+
+    fetchAndMatchServerIssues(virtualFile, engine, moduleKey, relativePath);
+  }
+
+  private void fetchAndMatchServerIssues(VirtualFile virtualFile, ConnectedSonarLintEngine engine, String moduleKey, String relativePath) {
+    // TODO make it possible to cancel
+    this.executorService.submit(() -> {
+      Iterator<ServerIssue> serverIssues = fetchServerIssues(engine, moduleKey, relativePath);
+
+      Collection<IssuePointer> serverIssuePointers = toStream(serverIssues).map(ServerIssuePointer::new).collect(Collectors.toList());
+
+      if (!serverIssuePointers.isEmpty()) {
+        store.storeServerIssues(virtualFile, serverIssuePointers);
+      }
+    });
+  }
+
+  private <T> Stream<T> toStream(Iterator<T> iterator) {
+    Iterable<T> iterable = () -> iterator;
+    return StreamSupport.stream(iterable.spliterator(), false);
+  }
+
+  private Iterator<ServerIssue> fetchServerIssues(ConnectedSonarLintEngine engine, String moduleKey, String relativePath) {
+    try {
+      return engine.downloadServerIssues(moduleKey, relativePath);
+    } catch (SonarLintWrappedException e) {
+      // download failed, fall back to local storage, if exists
+      return engine.getServerIssues(moduleKey, relativePath);
+    } catch (Throwable t) {
+      SonarLintConsole.get(myProject).error("could not get server issues", t);
+      return Collections.emptyIterator();
+    }
+  }
+
+  private String getRelativePath(VirtualFile virtualFile) {
+    if (myProject.getBasePath() == null) {
+      throw new IllegalStateException("no base path in default project");
+    }
+    return Paths.get(myProject.getBasePath()).relativize(Paths.get(virtualFile.getPath())).toString();
+  }
+
+  @Override
+  public void initComponent() {
+    this.executorService = Executors.newFixedThreadPool(THREADS_NUM);
+  }
+
+  @Override
+  public void disposeComponent() {
+    executorService.shutdown();
```

Comment:
It only rejects new tasks, so has no practical effect in this case. I would try shutdownNow, even with a bad support of interrupts by the thread